fastn-core/src/auth/emailpassword.rs

```diff
@@ -0,0 +1,371 @@
+        return fastn_core::error::Error::generic_err("no session found for the given session_id");
+    }
+
+    Ok(fastn_core::auth::FastnUser::from_row(
+        rows.first().expect("vec of length 1 must have first()"),
+    ))
+}
+
+// insert a row in fastn_session table
+pub async fn create_session(
+    session_id: &uuid::Uuid,
+    user_id: &uuid::Uuid,
+) -> fastn_core::Result<u64> {
+    let client = db::get_client().await?;
+
+    Ok(client
+        .execute(
+            "insert into fastn_session(id, user_id) values ($1, $2)",
+            &[&session_id, &user_id],
+        )
+        .await
+        .unwrap())
+}
+
+pub async fn destroy_session(session_id: uuid::Uuid) -> fastn_core::Result<u64> {
+    let client = db::get_client().await?;
+
+    Ok(client
+        .execute("delete from fastn_session where id = $1", &[&session_id])
+        .await
+        .unwrap())
+}
+
+pub async fn insert_oauth_token(
+    session_id: &uuid::Uuid,
+    token: &str,
+    provider: fastn_core::auth::AuthProviders,
+) -> fastn_core::Result<u64> {
+    let client = fastn_core::auth::emailpassword::db::get_client().await?;
+
+    let id = uuid::Uuid::new_v4();
+
+    Ok(client
+        .execute(
+            "insert into fastn_oauthtoken(id, session_id, token, provider) values ($1, $2, $3, $4)",
+            &[&id, &session_id, &token, &provider.as_str()],
+        )
+        .await
+        .unwrap())
+}
+
+pub async fn get_token_from_db(
+    session_id: &uuid::Uuid,
+    provider: &str,
+) -> fastn_core::Result<Token> {
+    let client = db::get_client().await?;
+
+    let token = client
+        .query(
+            "select session_id, token, provider from fastn_oauthtoken where session_id = $1 and provider = $2",
+            &[&session_id, &provider],
+        )
+        .await
+        .unwrap_or(vec![]);
+
+    if token.is_empty() {
+        return fastn_core::error::Error::generic_err("no token exists");
+    }
+
+    let token = token.first().expect("vec of length must have first");
+
+    Ok(Token::from_row(token))
+}
+
+pub async fn get_user_from_session(
+    session_id: &uuid::Uuid,
+) -> fastn_core::Result<fastn_core::auth::FastnUser> {
+    let client = db::get_client().await?;
+
+    let rows = client
+        .query(
+            "select id, username, name, email \
+                    from fastn_user \
+                    where id = (select user_id from fastn_session where id = $1)",
+            &[&session_id],
+        )
+        .await
+        .unwrap_or(vec![]);
+
+    if rows.is_empty() {
+        return fastn_core::error::Error::generic_err("no session found for the given session_id");
+    }
+
+    Ok(fastn_core::auth::FastnUser::from_row(
+        rows.first().expect("vec of length 1 must have first()"),
+    ))
+}
+
+// TODO: this is borrowed from pg.rs
+// pg.rs and this mod should use the same code
+pub mod db {
+    async fn create_pool() -> Result<deadpool_postgres::Pool, deadpool_postgres::CreatePoolError> {
```

Comment:
This a fastn_core global functionality, a lot of code may need db access tomorrow, let's move this to `fastn_core::db`.

Reply:
`fastn_core::db` exists now and used by the auth code. We'll incorporate using this in the `sql` processor in a separate pr.